felix/bpf/ut/bpf_prog_test.go

```diff
@@ -864,189 +893,164 @@ var udpDefault = &layers.UDP{
-			// protocol below.
-			ipv6 = new(layers.IPv6)
-			*ipv6 = *ipv6Default
+			_ = pkt.udp.SetNetworkLayerForChecksum(pkt.ipv4)
 		}
-	} else {
-		if ipv4Alt != nil {
-			ipv4 = ipv4Alt
+	case layers.IPProtocolTCP:
+		if pkt.isIPv6 {
+			_ = pkt.tcp.SetNetworkLayerForChecksum(pkt.ipv6)
 		} else {
-			// Make a copy so that we do not mangle the default if we set the
-			// protocol below.
-			ipv4 = new(layers.IPv4)
-			*ipv4 = *ipv4Default
+			_ = pkt.tcp.SetNetworkLayerForChecksum(pkt.ipv4)
 		}
 	}
+}
 
-	if l4Alt == nil {
-		l4Alt = udpDefault
-	}
-
-	if payloadAlt != nil {
-		payload = payloadAlt
-	} else {
-		payload = payloadDefault
+func (pkt *Packet) handleIPv6() {
+	if pkt.ipv6 == nil {
+		// Make a copy so that we do not mangle the default if we set the
+		// protocol below.
+		var ipv6 *layers.IPv6 = new(layers.IPv6)
+		*ipv6 = *ipv6Default
+		*pkt.ipv6 = *ipv6
 	}
-
-	return generatePacket(eth, ipv4, ipv6, l4Alt, payload, ipv6Enabled)
-
+	pkt.ipv6.NextHeader = pkt.protocol
+	pkt.ipv6.Length = uint16(pkt.length)
+	pkt.layers = append(pkt.layers, pkt.ipv6)
 }
 
-func generatePacket(eth *layers.Ethernet, ipv4 *layers.IPv4, ipv6 *layers.IPv6, l4 gopacket.Layer, payload []byte, ipv6Enabled bool) (
-	*layers.Ethernet, *layers.IPv4, *layers.IPv6, gopacket.Layer, []byte, []byte, error) {
-	var (
-		udp    *layers.UDP
-		tcp    *layers.TCP
-		icmp   *layers.ICMPv4
-		icmpv6 *layers.ICMPv6
-	)
-
-	if l4 != nil {
-		switch v := l4.(type) {
-		case *layers.UDP:
-			udp = v
-			if ipv6Enabled {
-				ipv6.NextHeader = layers.IPProtocolUDP
-			} else {
-				ipv4.Protocol = layers.IPProtocolUDP
-			}
-		case *layers.TCP:
-			tcp = v
-			if ipv6Enabled {
-				ipv6.NextHeader = layers.IPProtocolTCP
-			} else {
-				ipv4.Protocol = layers.IPProtocolTCP
-			}
-		case *layers.ICMPv4:
-			icmp = v
-			ipv4.Protocol = layers.IPProtocolICMPv4
-		case *layers.ICMPv6:
-			icmpv6 = v
-			ipv6.NextHeader = layers.IPProtocolICMPv6
-		default:
-			return nil, nil, nil, nil, nil, nil, errors.Errorf("unrecognized l4 layer type %t", l4)
-		}
+func (pkt *Packet) handleIPv4() {
+	if pkt.ipv4 == nil {
+		// Make a copy so that we do not mangle the default if we set the
+		// protocol below.
+		var ipv4 *layers.IPv4 = new(layers.IPv4)
+		*ipv4 = *ipv4Default
+		pkt.ipv4 = ipv4
 	}
+	pkt.ipv4.Protocol = pkt.protocol
+	pkt.ipv4.Length = uint16(pkt.length + 5*4)
+	pkt.layers = append(pkt.layers, pkt.ipv4)
+}
 
-	switch {
-	case udp != nil:
-		if ipv6Enabled {
-			ipv6.Length = uint16(8 + len(payload))
-			_ = udp.SetNetworkLayerForChecksum(ipv6)
-			udp.Length = uint16(8 + len(payload))
-
-			pkt := gopacket.NewSerializeBuffer()
-			err := gopacket.SerializeLayers(pkt, gopacket.SerializeOptions{ComputeChecksums: true},
-				eth, ipv6, udp, gopacket.Payload(payload))
-			return eth, nil, ipv6, udp, payload, pkt.Bytes(), err
+func (pkt *Packet) handleEthernet() {
+	if pkt.eth == nil {
+		if pkt.isIPv6 {
+			pkt.eth = ethDefaultWithIPv6
```

Comment:
similarly you could have `Packet.L2Proto()`  and set the appropriate proto in the eth header, the MACs are the same. And we could also check that the supplied eth proto isnt nonsense based on the L3, perhaps, but it is test ...